#164878049 Disable contextmenu on H mark image

Luleherll opened the pull request:
### PT Story: H-mark image: users are right-clicking to just save it; it's not the customized version
https://www.pivotaltracker.com/n/projects/1904891/stories/164878049

Changes proposed in this pull request:
1. Disable right-clicking on the H-mark
2. Add a custom context menu
3. Translate tooltip message
4. Change H-mark download buttons' color and layout
5. Write tests

Screenshots (Optional):

![Screen Shot 2019-09-12 at 7 36 35 PM](https://user-images.githubusercontent.com/36928255/64803077-b6303700-d594-11e9-930e-6b0655101459.png)

Ready for review:
@weedySeaDragon @thesuss @patmbolger

Luleherll committed:
add some code

Luleherll committed:
ft(H-mark image): disable contextmenu on the H-mark image

- disable contextmenu
- add a tooltip on the image

[Finishes #164878049]

weedySeaDragon reviewed (changes requested):
1. Change the title of the PR to start with "WIP" to show that this is a "Work In Progress" -- that more work will be coming after you get feedback.  I don't know if we discussed this; I likely didn't mention this to you.
   - need to get feedback from @thesuss and @patmbolger to see if they like this approach and if so, have suggestions for the specific wording in the tooltip
2. add these to your list of changes that will need to be included in this PR:
    1. cucumber test(s)
    2. change CSS so that this tooltip (`.tooltip-text `) has the same style as the others (`fas_tooltip(...)`)

patmbolger commented:
Initial comments (in addition to @weedySeaDragon 's):

1) The download test is on _right_ of the image (not the left)

2) The tooltip message has to be translated to Swedish (and make sure that is shown when the language is set to that)

3) I wonder if we should only show that tooltip _when the user right right-clicks_ on the image?  It might be _slightly_ less annoying that way, although it may not be worth worrying about.

4) I'm wondering if - instead of blocking the entire context menu, and using the tooltip - if we could just remove the menu items related to the image (save, download, get address, etc.).
   - if this is a lot of work (including browser-specific code) then it's not worth it,
   - as a developer, it would be nice to be able to access the "Inspect Element" option, but that is also not worth a lot of work.

4) If we're going to do this, it makes sense to also apply the same treatment to the proof-of-membership(?)

patmbolger reviewed (commented):
Pls see comments.

thesuss commented on app/views/companies/_company_h_brand.html.haml:
```diff
@@ -4,7 +4,8 @@
 
-    #company-h-brand
+    #company-h-brand{ "onContextMenu" => "return false;" }
+      .tooltip-text Please use the links on the left for your H-Brand download.
```
I guess this need to be t() ?

rmzse commented:
One thing about the text design is to make the purpose clearer by cleaning it up. Make the buttons stand out as clear CTA's should and make the text shorter. A good way to improve UX is to show users what they can do and why.

Changing the text to "Download Image" and "Copy Link to Image" is a good start, check out the image.

The idea is to have the design work for you so you don't have to explain so much, which in turn makes intructions shorter and more compact. Also, by removing the CTA's from to outside of the text paragraph it is easier for the user to really understand what to do instead of insisting to right-click the image.

A better UX would actually be to have the image rendered (once) and the setup of the business through ImageMagick.

<img width="1041" alt="Screenshot at aug  22 19-28-06" src="https://user-images.githubusercontent.com/13518870/63536259-8355ce80-c513-11e9-99f5-c701a326edf3.png">

Luleherll committed:
ft(H-markt): Change CTA buttons layout

- make context message show onclick only
- disable contextmenu on 'proof of membership' image
- Change CTA buttons layout

[#164878049]

Luleherll committed:
ft(H-markt): Change CTA buttons layout

- make context message show onclick only
- disable contextmenu on 'proof of membership' image
- Change CTA buttons layout

[#164878049]

Luleherll committed:
Git pull

Luleherll committed:
Add new line

weedySeaDragon commented on config/database.yml:
```diff
@@ -9,10 +9,14 @@ default: &default
 development:
   <<: *default
   database: shf_project_development
+  username: postgres
```
You still need to get this figured out.  Either don't commit it locally or figure out how to access your db with the default project credentials.

weedySeaDragon commented on config/locales/en.yml:
```diff
@@ -766,15 +766,15 @@ en:
-                        take a screenshot, or link to it from your website
-                        (%{show_link}), and use it in social media or print to use in a
-                        physical context, for example, to show a customer.
+      image_how_to_use_html: You can download the image,take a screenshot. 
```
Need to fix grammar and capitalization to:
```
image_how_to_use_html:  You can download the image, take a screenshot, link to it from you website and use it in social media, or print it have and show (ex: show it to a customer/client).
```

weedySeaDragon commented on config/locales/en.yml:
```diff
@@ -766,15 +766,15 @@ en:
       download_image: Download Image
-      show_image: Show Image
+      show_image: Copy link of the image
+      custom_context_text: Please use the links on the right for your H-Brand download.
```
Go ahead and add this same entry  to the `sv.yml` locale file, including the English words.  @thesuss tell you the Swedish for it later.

That will at least get the RSpec tests to pass (hopefully) for now.

weedySeaDragon commented on app/assets/stylesheets/company-h-brand.css:
```diff
@@ -14,30 +15,54 @@ body {
+  font-weight: bold;
+}
+
+.custom-menu {
```
The style for `.custom-menu` needs to be used with both the proof-of-membership _and_ the company-h-brand. This file is for styles that _only_ apply to the company-h-brand:  You can tell that by the naming convention: the name of this .css file matches a specific partial.

So pull this out into `user.scss` (= the styles for the `user` views).

If you have to (if the above doesn't work), you can put this out into it's own file (e.g. custom-menu.css) and then have it be included (via stylesheet_tags) in both the _proof-of-membership and _company-h-brand partials

weedySeaDragon reviewed (changes requested):
I have some thoughts on layout and style, but want to wait to hear the  feedback from @patmbolger and @thesuss about what they think of this in general.

Also -- I have the cucumber tests written and passing.  I did need to add a `cucumber step` to handle right-clicking.  Will post about that later.

Luleherll committed:
ft(H-markt): change download buttons color

- create custom-menu file
- fix custom context bug
- update yaml files

Luleherll committed:
ft(H-markt): writes tests for the functionality

- add tests
- add translation

weedySeaDragon commented on config/database.yml:
```diff
@@ -9,14 +9,10 @@ default: &default
 development:
   <<: *default
   database: shf_project_development
```
:tada:

weedySeaDragon commented on config/locales/en.yml:
```diff
@@ -766,9 +766,9 @@ en:
-                            Link to it from your website and use it in social media.
-                            Or Print to use in a physical context, for example, to show a customer.
+      image_how_to_use_html: You can download the image, take a screenshot,
+                            link to it from you website and use it in social media,
```
"you" should be "your"

weedySeaDragon commented on config/locales/sv.yml:
```diff
@@ -788,7 +788,8 @@ sv:
                                 Använd den här länken på en extern webbplats:
                                 </span>"
-      show_image: Visa bild
+      show_image: Kopiera bildadress
```
Don't change this one because it makes the meaning of the key (`show_image `) different from the meaning of the value (`copy the url`).

Instead, add a new entry (line) below (you'll need to ensure that the indentation is correct):

```
copy_image_url: Kopiera bildadress
```
Likewise for the English translation file (`en.yml`), add the same line but with the English words: `copy_image_url: Copy the URL`

_**Then**_ you need to be sure to use the `users.show.copy_image_url` translation key instead of the old one (old = `users.show.show_image`).
(That includes places in cucumber tests.)

weedySeaDragon commented on app/views/users/show.html.haml:
```diff
@@ -90,10 +90,10 @@
                                    proof_of_membership_path(@user, render_to: :jpg),
-                                   class: 'btn btn-sm btn-secondary')
+                                   class: 'btn btn-sm btn-primary')
         = link_to(t('.show_image'),
```
We now need to change the functionality:  The action of this button needs to (1) get the short url for the proof of membership, (2) copy it to the user's clipboard, and then (3) show a message to the user that the information has been copied to their clipboard and that they can now paste it where needed.

Just glancing quickly at the code (you should look more carefully at it to verify), seems that you need to use the `short_proof_of_membership_url(user)` method to get the url.

We'll need new text in the locale files for (3) above.  See my review comments in the `sv.yml` file.

weedySeaDragon replied:
We'll need the same stuff for the Company H-markt  (need to copy the URL to the clipboard instead of showing the image and add the entries to the locale files)

weedySeaDragon reviewed (changes requested):
Good work!  It's getting there.

Now it's clear enough for me to realize that we need to change the _functionality_ so that the button actually _copies the URL to the clipboard_ -- which is different from what it was doing before.  See my comments in the `app/views/users/show.html.haml` file.

Please also paste a new screenshot into the PR description so that we can see the latest version (since it's now been improved so much).

Luleherll committed:
Remove white space in parameter types to make tests pass

patmbolger commented:
Looking at the latest version of the PR:

1. I think we need a little space after the blue buttons - set the bottom margin the same as the top margin.

<img width="380" alt="Screen Shot 2019-09-05 at 10 58 30 AM" src="https://user-images.githubusercontent.com/9968213/64353712-33274380-cfcc-11e9-8c64-1e9801f28f18.png">

2. When I click "download image", this is what I get:

<img width="288" alt="Screen Shot 2019-09-05 at 11 00 17 AM" src="https://user-images.githubusercontent.com/9968213/64353813-5e119780-cfcc-11e9-8bef-69b5c17066b5.png">

It shouldn't have that text at the top.

3. "Copy link of the image" does not actually do that - it goes to a page where the user can see how the image renders, get it's size, and also get a shortened URL if they want to use that on the own website.  So, either we change the wording ("View Image", or similar), or we change the action.

We should discuss at our next meeting and get consensus on how we want this to work.

Luleherll committed:
create custom context partial

Luleherll committed:
ft(h-markt): implement custom contextmenu

- implement copy_to_clipboard button
- write logic for contextmenu actions
- update test

[Finishes #164878049]

Luleherll committed:
Change  swedish translation and add bottom margin on image action buttons

patmbolger commented on app/assets/javascripts/users.js:
```diff
@@ -1,17 +1,63 @@
+    $("#editStatusModal").modal("hide");
+  });
+
+  $(".custom-context").bind("contextmenu", e => {
```
Please note that jQuery `.bind` has been deprecated as of jQuery 3.0 (we're at 3.4.1).  We should use `.on1 instead.

patmbolger reviewed (changes requested):
The code looks very nice!  I have just one remaining comment - other than that I am fine with the PR.

Luleherll committed:
change  to
